geopandas/tools/make_grid.py

```diff
@@ -0,0 +1,387 @@
+            x_dist = bounds[3] - bounds[1] - offset[1]
+
+        # Create rectangular meshgrid containing both the centers and the cornes
+        # of the hexagonal grid
+        dx = cell_size
+        dy = cell_size * np.sqrt(3) / 2
+
+        # Determine number of grid points along x/y direction such that the area of
+        # any bounding box is fully covered with polygons.
+        n_dx = x_dist / dx
+        n_grid_point_x = int((np.ceil((n_dx - 1) / 1.5) + 2) * 1.5)
+        n_dy = y_dist / dy
+        n_grid_point_y = int(n_dy) + 4
+
+        x_coords = np.arange(0, n_grid_point_x).astype(float) * dx
+        y_coords = np.arange(0, n_grid_point_y).astype(float) * dy - dy
+        xv, yv = np.meshgrid(x_coords, y_coords)
+
+        # Shift every second row to transform the rectangular into a hexagonal grid
+        xv[::2, :] = xv[::2, :] - (dx / 2)
+
+        mask_center = np.zeros_like(xv, dtype=bool)
+        mask_center[1::2, 2::3] = True
+        mask_center[::2, 1::3] = True
+
+        if what == "centers":
+            hex_centers_np = np.array([xv[mask_center], yv[mask_center]]).T.reshape(
+                -1, 2
+            )
+            if not flat_topped:
+                hex_centers_np = hex_centers_np[:, ::-1]
+
+            output_grid = points_from_xy(
+                hex_centers_np[:, 0] + grid_origin_x,
+                hex_centers_np[:, 1] + grid_origin_y,
+            )
+
+        elif what == "corners":
+            # The inverted center mask is the corner mask. Now consider all corners
+            mask_corners = np.invert(mask_center)
+
+            hex_corners_np = np.array([xv[mask_corners], yv[mask_corners]]).T.reshape(
+                -1, 2
+            )
+
+            if not flat_topped:
+                hex_corners_np = hex_corners_np[:, ::-1]
+
+            output_grid = points_from_xy(
+                hex_corners_np[:, 0] + grid_origin_x,
+                hex_corners_np[:, 1] + grid_origin_y,
+            )
+
+        elif what == "polygons":
+            hex_coords_a = _hex_polygon_corners(xv, yv, (0, 1))
+            hex_coords_b = _hex_polygon_corners(xv, yv, (2, 0))
+
+            hex_coords = np.concatenate((hex_coords_a, hex_coords_b), axis=0)
+
+            if not flat_topped:
+                hex_coords = hex_coords[:, :, ::-1]
+
+            hex_coords[:, :, 0] += grid_origin_x
+            hex_coords[:, :, 1] += grid_origin_y
+
+            hex_polygons = [Polygon(hex_set) for hex_set in hex_coords]
+
+            hex_polygons_np = np.array(hex_polygons)
+            output_grid = from_shapely(hex_polygons_np)
+
+        else:
+            raise ValueError("Invalid value for parameter `what`")
+    else:
+        raise ValueError("Invalid value for parameter `cell_type`")
+
+    output_grid = GeoSeries(output_grid)
+
+    if isinstance(input_geometry, (GeoDataFrame, GeoSeries)):
+        if input_geometry.crs is not None:
+            output_grid = output_grid.set_crs(input_geometry.crs)
```

Comment:
```suggestion
    output_grid = GeoSeries(output_grid, crs=input_geometry.crs)
```
This will have the same effect.

Reply:
I added the is instance check because input_geometry could also be a shapely polygon.